Tabs .getCurrent, .onTabSelectionChanged, .getAllTabs for Chrome and FF

ssawchenko opened the pull request:
This branch now contains the changes for all three tab functions in both Chrome and FF.

ssawchenko committed:
Add support for local button icons

* Use data.url to patch up local icon path

ssawchenko committed:
Added tabs getCurrent and onTabSelectionChanged

* Note, getCurrent appears to work in Chrome and FF; IE
  implementation still required
* onTabSelection appears to work in Chrome, but not FF; IE
  implementation still required
* Reverted last change for setIcon

ssawchenko committed:
Further fixes for onTabSelectionChanged

* onTabSelection appears to work in Chrome, but FF requires that
  the method be called each time the user wishes to listen for the next
  tab change; IE implementation still required

ssawchenko committed:
Added tabs.getAllTabs for Chrome and FF

antoinevg committed:
Support Firefox 33 with Add-on SDK 1.17

reicolina commented on chrome/assets_forge/api-priv-chrome.js:
```diff
@@ -236,6 +236,51 @@ var apiImpl = {
+  						success({id: tab.id, url: tab.url});
+  					}
+  					else {
+  						// TODO error
```
@ssawchenko I would like this 'todo' to be implemented before merging to master.

reicolina commented on chrome/assets_forge/api-priv-chrome.js:
```diff
@@ -236,6 +236,51 @@ var apiImpl = {
+	  						success({id: tab.id, url: tab.url});
+	  					}
+	  					else {
+	  						// TODO error
```
@ssawchenko I would like this 'todo' to be implemented before merging to master.

reicolina commented on firefox/template-app/lib/main.js:
```diff
@@ -252,6 +252,33 @@ var apiImpl = {
+		 * to get the next tab selection, it must call this function again.
+		 *   TODO Figure out why success can only be called once below (subsquent calls to success appear to fail, 
+		 *   so I am guessing that the framework is doing something to the function reference after it is called)
+		 */
```
@ssawchenko just courious...were you able to figure this one out?

reicolina committed:
Merge pull request #7 from trigger-corp/master

Support Firefox 33 with Add-on SDK 1.17

Commit:
Fixed for...in loop syntax in custom Firefox code.

* Was breaking the extension when this code is minified (i.e. when forge-extension build is run without the -d debug switch).

ssawchenko committed:
Merge remote-tracking branch 'origin/from-original' into shayla

Commit:
Changed getAllTabs() to iterate over the sdk/tabs object prototype.

*  It turned out that the Mozilla sdk/tabs object was not suitable for use in a for...in loop and the technique recommended by Mozilla (using the ECMAScript 6 for...of construct) broke OpenForge minification.

antoinevg committed:
Use original content script proxy for Firefox v32 or lower

reicolina committed:
Merge pull request #8 from trigger-corp/master

Use original content script proxy for Firefox v32 or lower

ssawchenko committed:
Merge branch 'from-original' into shayla

Commit:
Partial implementation of new prefs.setSync, prefs.getSync and prefs.clearSync methods.

Commit:
Fix to the internal calling of the new prefs.getSync method.

Commit:
PEUS-757: further fixes for the new synchronous prefs methods in IE.

Commit:
Further fixes for IE.

mcnulty committed:
Fix setting of active tab when frame process is 64-bit, but tab processes are 32-bit.

- the difference in bitness requires use of the HWNDX type.
- from what I can tell, the default behavior of Internet Explorer 11 is
  to run the frame process under 64-bit and the tab processes as 32-bit
  for compatibility with older plugins. If Enhanced Protection mode is
  enabled (which requires a restart), the tab processes run as 64-bit.

mcnulty committed:
Fix forge.button.setIcon on IE.

- in FrameServer::ProxyListen, the currentToolbar variable (an index
  into FrameServer.m_buttons) was being set before the m_activeToolbar
  member was initialized (or modified) by the LoadCommand. The
  currentToolbar is now being set on every time through the channel
  command processing loop.
- Also, made button lookup by the toolbar a bit more robust by using
  hash_map.find instead of operator[]. operator[] will create an
  "empty" button when the toolbar is invalid/unset/etc and this was
  causing button_setIconCommand to either fail or set the icon for
  incorrect button (such as the Home button on IE 10 or 11).

mcnulty committed:
Update IE native code.

Commit:
Further fixes for getSync, setSync and clearSync in IE.

Commit:
Added support for storing strings with forge.db.* methods.

mcnulty committed:
Update IE binaries.

mcnulty committed:
Add native IE function to close the popup.

- refactors CNativeControls::popup_visible to use its second parameter
  to determine whether to show or hide the popup -- currently, this
  function was only being used to show the popup.

ssawchenko committed:
package.json now correctly reads uuid

Commit:
PEUS-757: fixes for prefs.getSync.

* IE and FF return null instead of undefined if preference is empty or doesn't exist
* change to all browsers to attempt to parse the prefs.getSync return value

mcnulty committed:
PEUS-685: IE C++ support for forge.tabs.onTabSelectionChanged.

- this functionality can be invoked from javascript with something like:
  window.messaging.active_tab_listen(forge.config.uuid,
            function(tabInfo) {
                if (typeof success !== "function") return;
                success({ id: tabInfo.id, url: tabInfo.url });
            });

Commit:
Added JavaScript support for new button.closePopup() method (IE only).

* renamed internal C++ close_popup() to button_closePopup() to match other forge button methods
* updated native code for IE

Commit:
PEUS-685: JavaScript support for forge.tabs.onTabSelectionChanged in IE.

mcnulty committed:
PEUS-684: add IE C++ support for forge.tabs.getAllTabs

- this commit includes modifications to the private IE javascript to
  demarshal the array of tab objects from a COM SAFEARRAY to a
  javascript array, using the IE specific VBArray type.

Commit:
PEUS-696: updated IE binaries.

ssawchenko committed:
prefs.db.get and prefs.async.db.get fix

* Fixed issue where db value containing all numbers is returned as a number
  instead of a string.

ssawchenko committed:
Fixed missing case on last commit

mcnulty committed:
Fix removal of correct RegKey for preferences in uninstaller.

Commit:
PEUS-696: added JS support for tabs.getCurrent in IE.

mcnulty committed:
IE BHO: Improve robustness of DOM attachment and script/style injection.

- page refresh events were not causing scripts/styles and DOM attachment
  to occur. This commit makes the logic more robust to add this support.
- 3 events from the WebBrowser2 object now can trigger DOM attachment
  and script/style injection: NavigateComplete2, DocumentComplete and
  DownloadComplete.
- DOM attachment and script/styles are now idempotent -- once the COM
  servers and scripts/styles are injected into a window DOM object,
  these operations will not occur again until the window DOM object
  is reloaded or the browser navigates to a new page.
- Idempotency is achieved by inspecting the propertys on the IHTMLWindow
  object -- after script/style injection, a property is added to the
  window to record the completion of this operation.

Commit:
PEUS-696: updated IE binaries following fixes to script injection in C++ code.

ssawchenko committed:
Fixed popup panel sizing on FF

* Added default_width and default_height button config params
* If above params included in config.json then the Panel width and height
  will be set to those values.

ssawchenko committed:
getAllTabs updated to use newer chrome.tabs.query method

ssawchenko committed:
Added missing files from last check in regarding tab methods

mcnulty committed:
Further improvements to script/style injection for IE.

- scripts/styles are not injected for the OnNavigateComplete2 event --
  the document object was not available for use by the scripts at this
  point in the page loading. Only the attachments to the window can
  occur at this point.
- COM servers are attached whenever various events occur -- some times
  the COM servers were not available even though the window properties
  for the servers existed.

Commit:
Updated IE binaries with latest script/style injection changes.

mcnulty committed:
IE C++: treat NULL BSTR as empty string ("")

- This is compliant with the expected semantics of BSTR:
  http://blogs.msdn.com/b/ericlippert/archive/2003/09/12/52976.aspx
- This was exposed by a recent update to IE11 (11.0.14), which perhaps also
  introduced updates to the JScript COM bridge where empty strings in JS
  are now passed via COM as NULL BSTRs.

Commit:
Updated IE binaries with the "treat NULL BSTR as empty string" fix for IE11 (11.0.14).

Commit:
Replaced forge internal XHR call with call to jQuery.ajax.

* fixes failing Ajax GET requests in IE8

Commit:
Restrict earlier jQuery.ajax() change in the IE code to target only IE8.

* IE8 will use jQuery.ajax() and newer versions of IE will use the custom forge AJAX implementation
* forge is currently bundled with jQuery version 1.5.2 and the AJAX implementation in this version doesn't work in IE11

Commit:
Set custom minimum popup dimensions for IE.

*  a future improvement could be to read these values from config.json button->default_width and button->default_height

ssawchenko committed:
PEUS-842: FF now clears data on uninstall/disable

* Added exports.onUnload to listen for unloading events.
* Due to known mozilla bug, uninstall emits a "disabled" event. This means
  that both disabled and uninstall are treated the same.
* Clear simple storage when uninstall/disabled occurs.

mcnulty committed:
IE C++: listen for document.onfocusin to catch browser window focus events.

- previously when moving between browser windows in IE, no events would
  be fired and the active tab would not be updated in the COM servers,
  thus making FG <=> BG messages non-functional.
- this commit installs an event handler for the document.onfocusin event
  that catches incoming focus events and updates the active tab
  accordingly.
- also, care needs to be taken to detach the previous document before
  adding the listener to the new document since ATL will assert when
  attempting to DispEventAdvise on a new document before unadvising.

Commit:
Updated native IE files following C++ fix to listen for document.onfocusin to catch browser window focus events.

Commit:
Make prefs.get return null in Firefox (instead of undefined) if the preference has not been set.

* bug reference PEUS-873

Commit:
Don't extend the Array prototype on IE pages running in older quirks mode.

* avoids errors encountered when adding indexOf support to the Array data type on pages with for example an HTML4 doctype declaration
* internal bug reference: PEUS-880

Commit:
Fix for message dispatch from Safari popover.

* removed JSON.stringify() of reply as internal.priv.receive() expects a parsed object

Commit:
Added support for custom "default_width" and "default_height" settings for button popover.

* added Python code to dynamically populate the button popover width and height fields of the Safari extension Info.plist file when "default_width" and "default_height" values are set in config.json

Commit:
Changed data type of custom "default_width" and "default_height" settings in build script.

* amended Python code to set the button popover width and height fields as type real instead of type string
* fixes retrieval of these custom properties when the extension is first built with Safari Extension Builder

mcnulty committed:
ID-199: IE C++: correctly convert UTF-8 XHR response into UTF-16 for transport over COM bridge

The HTTP response handling for the XHR request assumed that the response
was always ASCII when converting the raw bytes to a wstring with:

wstring response(this->bytes.begin(), this->bytes.end());

For multi-byte characters, this conversion is incorrect.

This fix converts the stream of bytes into a wstring using the
appropriate conversion function.

Commit:
ID-199: updated IE binaries with UTF-8 to UTF-16 conversion fix.

mcnulty committed:
PINGONESTG-863: IE C++: remove errant TODO about memory leak

NativeExtensions::getURL uses an output parameter for the URL.
Per COM Memory Management Rules ([1]), the caller will handle
de-allocating this data so this is not a memory leak as the
TODO indicated.

[1]: https://msdn.microsoft.com/en-us/library/windows/desktop/ms686638(v=vs.85).aspx

Commit:
ID-148: increased popup height to 270px to accommodate new beta mode control.

Commit:
PINGONESTG-863: only append log messages to the background console element when logging is enabled.

* in a production setting (with logging disabled), this background logging function was being frequently called - incessantly adding new <span> elements to the background console DOM element and creating a memory leak.

Commit:
Merge branch 'ID-141' into 'shayla'

* ID-148: increased popup height to 270px to accommodate new beta mode control.

Commit:
ID-369: increased popup width to accommodate longer content translations.

Commit:
Added PDB symbols information for DLLs from the latest build.
